test/hotspot/jtreg/compiler/vectorization/TestRoundVectorDoubleRandom.java

```diff
@@ -0,0 +1,148 @@
+ *
+ * @library /test/lib /
+ * @run driver compiler.vectorization.TestRoundVectorDoubleRandom
+ */
+
+package compiler.vectorization;
+
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+import compiler.lib.ir_framework.DontCompile;
+import compiler.lib.ir_framework.IR;
+import compiler.lib.ir_framework.IRNode;
+import compiler.lib.ir_framework.Run;
+import compiler.lib.ir_framework.RunInfo;
+import compiler.lib.ir_framework.Test;
+import compiler.lib.ir_framework.TestFramework;
+import compiler.lib.ir_framework.Warmup;
+
+public class TestRoundVectorDoubleRandom {
+  private static final int ITERS  = 11000;
+  private static final int ARRLEN = 997;
+  private static final double ADD_INIT = -7500.;
+
+  private static final double[] input = new double[ARRLEN];
+  private static final long [] res = new long[ARRLEN];
+  private static final Random rand = new Random();
+
+  public static void main(String args[]) {
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=8");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=16");
+    TestFramework.runWithFlags("-XX:-TieredCompilation", "-XX:CompileThresholdScaling=0.3", "-XX:MaxVectorSize=32");
+  }
+
+  @DontCompile
+  long golden_round(double d) {
+    return Math.round(d);
+  }
+
+  @Test
+  @IR(counts = {IRNode.ROUND_VD, "> 0"})
+  void test_round(long[] a0, double[] a1) {
+    for (int i = 0; i < a0.length; i+=1) {
+      a0[i] = Math.round(a1[i]);
+    }
+  }
+
+  @Run(test = "test_round")
+  @Warmup(ITERS)
+  void test_rounds(RunInfo runInfo) {
+    // Initialize
+    for (int i = 0; i < ARRLEN; i++) {
+      double val = ADD_INIT+(double)i;
+      input[i] = val;
+    }
+
+    test_round(res, input);
+    // skip test/verify when warming up
+    if (runInfo.isWarmUp()) {
+      return;
+    }
+
+    int errn = 0;
+    final int e_shift = 52;
+    final int e_width = 11;
+    final int e_bound = 1 << e_width;
+    final int f_width = e_shift;
+    final long f_bound = 1 << f_width;
+    final int f_num = 256;
+
+    // prepare test data
+    long fis[] = new long[f_num];
+    int fidx = 0;
+    for (; fidx < f_width; fidx++) {
+      fis[fidx] = 1 << fidx;
+    }
+    fis[fidx++] = 0;
+    for (; fidx < f_num; fidx++) {
+      fis[fidx] = ThreadLocalRandom.current().nextLong(f_bound);
+    }
+
+    // run test & verify
+    for (long fi : fis) {
+      final int e_start = rand.nextInt(9);
+      final int e_step = (1 << 3) + rand.nextInt(3);
+      for (int ei = e_start; ei < e_bound; ei += e_step) {
+        int ei_idx = ei/e_step;
+        long bits = (ei << e_shift) + fi;
+        input[ei_idx*2] = Double.longBitsToDouble(bits);
+        bits = bits | (1 << 63);
+        input[ei_idx*2+1] = Double.longBitsToDouble(bits);
+      }
```

Comment:
It's testing following cases:
1. all the `e` range, e.g. for double it's 11 bits, for float it's 8 bits
2. for `f` I add a special value `0` explicitly `fis[fidx++] = 0;`
3. for sign, both `+` and `-` are tested.

So, yes, it will test cases like `+/- 0, infty, NaN`.

Reply:
Can you refactor or at least comment the code a little better, or use more expressive variable names?
I'd have to spend a bit of time to understand your generation method here, and if I think that it is exhaustive and covers the special cases with enough frequency.

Reply:
Sure, I will add some comments to illustrate how it works.